proposed/middleware.md

```diff
@@ -0,0 +1,173 @@
+with other middleware components, in the processing of an incoming HTTP Request
+and the creation of a resulting HTTP Response, as defined by PSR-7.
+
+Middleware components MUST implement this interface.
+
+Middleware consumers (e.g. frameworks and middleware stacks) MUST type-hint any
+method accepting middleware components as arguments formally as `callable`, and
+informally as `Psr\Http\Middleware\MiddlewareInterface`, e.g. using php-doc tags:
+
+```php
+/**
+ * @param MiddlewareInterface $middleware
+ */
+public function push(callable $middleware)
+{
+    // ...
+}
+```
+
+## 2. Interfaces
+
+```php
+namespace Psr\Http\Middleware;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+interface MiddlewareInterface
+{
+    /**
+     * Process a request and return a response.
+     *
+     * Takes the incoming request and optionally modifies it before delegating
+     * to the next handler to get a response. May modify the response before
+     * ultimately returning it.
+     *
+     * @param RequestInterface $request
+     * @param ResponseInterface $response
+     * @param callable $next
+     *
+     * @return ResponseInterface
+     */
+    public function __invoke(
+        RequestInterface $request,
+        ResponseInterface $response,
+        callable $next
+    );
+}
+```
+
+### 2.1 Server vs Client Requests
+
+Some middleware components may require the additional methods defined by
+`ServerRequestInterface`, such as reading cookies or setting attributes.
+In this case, the middleware MUST check the type of the request before
+using these methods.
+
+To make this experience consistent an exception and trait are included in this package.
+
+Middleware components SHOULD use this exception when a `ServerRequestInterface`
+is expected and not provided:
+
+```php
+namespace Psr\Http\Middleware;
+
+use InvalidArgumentException;
+
+class ServerRequestRequiredException extends InvalidArgumentException
+{
+    /**
+     * Create an exception for a middleware that requires a server request.
+     *
+     * @param object $middleware
+     *
+     * @return static
+     */
+    public static function requiredBy($middleware)
+    {
+        return new static(sprintf(
+            'Middleware `%s` requires that the request be an implementation of `ServerRequestInterface`',
+            get_class($middleware)
+        ));
+    }
+}
+```
+
+Middleware components MAY use this trait to implement checking for `ServerRequestInterface`:
+
+```php
+namespace Psr\Http\Middleware;
+
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+trait RequiresServerRequestTrait
+{
+    /**
+     * Check that the given request is a server request.
+     *
+     * @param RequestInterface $request
+     *
+     * @return void
+     *
+     * @throws ServerRequestRequiredException
+     */
+    private function assertServerRequest(RequestInterface $request)
```

Comment:
IMO it makes it part of the standard, while it isn't. The less content the better it is. And it's an already existing pattern within the FIG.